goblet/deploy.py

```diff
@@ -1,271 +1,303 @@
-                parent_key="resource",
-                parent_schema=self.run_name,
-                params={"body": policy_bindings},
-            )
-
-    def _cloudfunction_delta(self, client, filename):
-        """Compares md5 hash between local zipfile and cloudfunction already deployed"""
-        self.zipf.close()
-        with open(filename, "rb") as fh:
-            local_checksum = base64.b64encode(checksum(fh, hashlib.md5())).decode(
-                "ascii"
-            )
-
-        source_info = client.execute(
-            "generateDownloadUrl", parent_key="name", parent_schema=self.func_name
-        )
-        resp = request("HEAD", source_info["downloadUrl"])
-        deployed_checksum = resp.headers["x-goog-hash"].split(",")[-1].split("=", 1)[-1]
-        modified = deployed_checksum != local_checksum
-        return modified
-
-    def _upload_zip(self, client):
-        """Uploads zipped cloudfunction using generateUploadUrl endpoint"""
-        self.zipf.close()
-        zip_size = os.stat(f".goblet/{self.name}.zip").st_size
-        with open(f".goblet/{self.name}.zip", "rb") as f:
-            resp = client.execute("generateUploadUrl", params={"body": {}})
-
-            requests.put(
-                resp["uploadUrl"],
-                data=f,
-                headers={
-                    "content-type": "application/zip",
-                    "Content-Length": str(zip_size),
-                    "x-goog-content-length-range": "0,104857600",
-                },
-            )
-
-        log.info("function code uploaded")
-
-        return resp["uploadUrl"]
-
-    def create_zip(self):
-        """Creates initial goblet zipfile"""
-        if not os.path.isdir(get_g_dir()):
-            os.mkdir(get_g_dir())
-        return zipfile.ZipFile(
-            get_g_dir() + f"/{self.name}.zip", "w", zipfile.ZIP_DEFLATED
-        )
-
-    def zip(self):
-        """Zips requirements.txt, python files and any additional files based on config.customFiles"""
-        config = GConfig()
-        self.zip_file("requirements.txt")
-        if config.main_file:
-            self.zip_file(config.main_file, "main.py")
-        include = config.customFiles or []
-        include.append("*.py")
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore")
-            self.zip_directory(get_dir() + "/*", include=include)
-
-    def zip_file(self, filename, arcname=None):
-        self.zipf.write(filename, arcname)
-
-    def zip_directory(
-        self,
-        dir,
-        include=["*.py"],
-        exclude=["build", "docs", "examples", "test", "tests", "venv"],
-    ):
-        exclusion_set = set(exclude)
-        globbed_files = []
-        for pattern in include:
-            globbed_files.extend(Path("").rglob(pattern))
-        for path in globbed_files:
-            if not set(path.parts).intersection(exclusion_set):
-                self.zipf.write(str(path))
+# from pathlib import Path
```

Comment:
we should still have core functionality of the deployer here

Reply:
Ok, so should the deployer call the backend's methods?

Reply:
Alternatively, could we turn the deployer into something like `common_cloud_actions.py` with a bunch of helper functions?